crates/red_knot_python_semantic/src/types/builder.rs

```diff
@@ -80,6 +82,17 @@ impl<'db> UnionBuilder<'db> {
                         return self;
                     } else if element.is_subtype_of(self.db, ty) {
                         to_remove.push(index);
+                    } else if ty_negated.is_subtype_of(self.db, *element) {
```

Comment:
Is there an argument for inlining the negation to ensure we avoid doing it if one of the above cases matches? We only use `ty_negated` once.

```suggestion
                    } else if ty.negate(self.db).is_subtype_of(self.db, *element) {
```

Reply:
> We only use `ty_negated` once.

It's still a loop though, so we would compute it for every pre-existing element of the union, no?

Reply:
Oh, I totally missed that, you're right. Let's leave it as is. If this shows up as a significant cost, I think there would be ways we can make it more efficient. For example, we could have a dedicated `negation_is_subtype_of` that would short circuit the most costly cases (where `ty` is a simple type and `ty_negated` requires constructing an intersection, in which case we will never be able to declare the resulting negation type a subtype of anything other than another negation type) and defer to `self.negate(db).is_subtype_of(...)` for the remainder.